### PR TITLE
feat: log c plus transformer

### PR DESCRIPTION
**Pull request by TremaMiguel**
This PR closes issue #274 

## Describe Your Changes

New transformer [LogCpTransformer](https://github.com/TremaMiguel/feature_engine/blob/feat_logc_transformer/feature_engine/transformation/log.py#L169)
 
As suggested by @GLevV, the `C` parameter can take an integer value or an `"auto"` option, in the second case the absolute value of the min is stored as `C`

```python
self.C = X[self.variables_].min(axis=0).abs() + 1
```

## How has this been tested

Same tests as Log Transformer added. [Link](https://github.com/TremaMiguel/feature_engine/blob/feat_logc_transformer/tests/test_transformation/test_logcp_transformer.py#L36)

## Examples

[LogCpTransformer](https://github.com/TremaMiguel/feature_engine/blob/feat_logc_transformer/docs/transformation/LogCpTransformer.rst) documentation includes example.

log_cp_example

### CODE REVIEW

File: tests/test_transformation/test_logcp_transformer.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+@pytest.mark.parametrize("log_base, exp_age, exp_marks", _test_automatic_find_variables)
+def test_logcp_base_plus_automatically_find_variables(
+    log_base, exp_age, exp_marks, df_vartypes
+):
+    transformer = LogCpTransformer(base=log_base, variables=None)
+    X = transformer.fit_transform(df_vartypes)
+
+    # expected output
+    transf_df = df_vartypes.copy()
```
Comment (solegalli) [inline tests/test_transformation/test_logcp_transformer.py:44]: The test file is a masterpiece. Thank you so much. 

Could we add tests for init param C and fit attribute C_?

either as part of tests 1 or 2 or separately if you find it better. Thank you!

File: docs/transformation/LogTransformer.rst [hunk 2]
```diff
@@ -21,7 +21,7 @@ Example
 	from feature_engine import transformation as vt
 
 	# Load dataset
-	data = data = pd.read_csv('houseprice.csv')
+	data = pd.read_csv('houseprice.csv')
```
Comment (solegalli) [inline docs/transformation/LogTransformer.rst:24]: thank you!

File: docs/transformation/LogCpTransformer.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+.. autoclass:: feature_engine.transformation.LogCpTransformer
+    :members:
+
+
+Example
+-------
+
+.. code:: python
+
+	import pandas as pd
```
Comment (solegalli) [inline docs/transformation/LogCpTransformer.rst:16]: indentation

File: docs/transformation/LogCpTransformer.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Load dataset
+    X, y = load_boston(return_X_y=True)
+    X = pd.DataFrame(X)
+
+    # Separate into train and test sets
+    X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size=0.3, random_state=0)
+
+    # set up the variable transformer
+    tf = vt.LogCpTransformer(variables = [7, 12])
```
Comment (solegalli) [inline docs/transformation/LogCpTransformer.rst:31]: not necessary but helpful if we do show param C in the init

File: docs/transformation/LogCpTransformer.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    # transform the data
+    train_t= tf.transform(X_train)
+    test_t= tf.transform(X_test)
+
+	# un-transformed variable
+    X_train[12].hist()
+
+.. image:: ../images/logcpraw.png
+
```
Comment (solegalli) [inline docs/transformation/LogCpTransformer.rst:44]: could we also display the output of C_? particularly useful when C is auto.

**Review by solegalli — COMMENTED** [review]
Hi @TremaMiguel 

This PR is really good. Thank you so much! 

Aside of the comments below, could you consider the following:

- [ ] could we add the new transformer to the list to be checked by sklearn's `check_estimator ` [here](https://github.com/solegalli/feature_engine/blob/master/tests/test_transformation/test_check_estimator_transformers.py) and make sure tests pass? They will after we change the C to C_
- [ ] In PR #273 we are adding `inverse_transform` to all transformers from this module. Would you mind adding it to this one as part of this PR you are currently working with?
- [ ] if you have time and don't mind, would you be able to add a jupyter notebook with an example to the examples folder? not necessary to merge this PR though.

File: feature_engine/transformation/log.py [hunk 6]
```diff
@@ -164,3 +164,161 @@ def _more_tags(self):
...
+    variables_:
+        The group of variables that will be transformed.
+
+    n_features_in_:
+        The number of features in the train set used in fit.
+
+    Methods
+    -------
+    fit:
+        This transformer does not learn parameters.
```
[the remark this answers is not in the document]
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:209]: Sure, will be something like 
```
This transformer learns the constant C to add to log transformation.
```

File: feature_engine/transformation/log.py [hunk 7]
```diff
@@ -164,3 +164,161 @@ def _more_tags(self):
...
+        if not isinstance(C, int) and not C == "auto":
+            raise ValueError("C can take only 'auto' or int")
+
+        self.variables = _check_input_parameter_variables(variables)
+        self.base = base
+        self.C = C
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        This transformer does not learn parameters.
```
[the remark this answers is not in the document]
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:235]: got it

File: feature_engine/transformation/log.py [hunk 8]
```diff
@@ -164,3 +164,161 @@ def _more_tags(self):
...
+
+        Select the numerical variables and determines whether the logarithm
+        can be applied on the selected variables (it checks if the variables
+        are all positive).
```
[the remark this answers is not in the document]
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:239]: got it

File: feature_engine/transformation/log.py [hunk 9]
```diff
@@ -164,3 +164,161 @@ def _more_tags(self):
...
+        -------
+        self
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+
+        # calculate C to add to each variable
+        if self.C == "auto":
+            self.C = X[self.variables_].min(axis=0).abs() + 1
```
[the remark this answers is not in the document]
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:270]: Sure, I'll change that. Besides, I've frequently seen the use of trailing underscores, but I'm not all familiar with it, could you share me a reference to better understand it ?

File: feature_engine/transformation/log.py [hunk 10]
```diff
@@ -164,3 +164,161 @@ def _more_tags(self):
...
+        The constant C to apply the transfomation log(x + C).
+
+        - If int, then log(x + C)
+        - If "auto", then C = abs(min(x)) + 1
+
+    Attributes
+    ----------
+    variables_:
+        The group of variables that will be transformed.
+
```
[the remark this answers is not in the document]
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:202]: got it

File: tests/test_transformation/test_logcp_transformer.py (hunk 1 above)
Comment (TremaMiguel) [reply tests/test_transformation/test_logcp_transformer.py:44]: got it

File: docs/transformation/LogCpTransformer.rst (hunk 4 above)
Comment (TremaMiguel) [reply docs/transformation/LogCpTransformer.rst:31]: got it

File: docs/transformation/LogCpTransformer.rst (hunk 5 above)
Comment (TremaMiguel) [reply docs/transformation/LogCpTransformer.rst:51]: got it

File: docs/transformation/LogCpTransformer.rst (hunk 3 above)
Comment (TremaMiguel) [reply docs/transformation/LogCpTransformer.rst:16]: not sure why locally this indentation doesn't appear 

```python
.. code:: python
    
    import pandas as pd
    import matplotlib.pyplot as plt
```

Comment (TremaMiguel) [reply docs/transformation/LogCpTransformer.rst]: fixed

**Comment by TremaMiguel** [comment]
@solegalli please take a second look, I have addressed all your comments.

File: feature_engine/transformation/log.py [hunk 11]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+    n_features_in_:
+        The number of features in the train set used in fit.
+
+    C_:
```
Comment (solegalli) [inline feature_engine/transformation/log.py:206]: could we put this attribute above n_features_in_ please?

File: feature_engine/transformation/log.py [hunk 12]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+        if not isinstance(C, int) and not C == "auto":
+            raise ValueError("C can take only 'auto' or int")
+
+        self.variables = _check_input_parameter_variables(variables)
+        self.base = base
+        self.C = C
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        This transformer learns the constant C to add to log transformation.
```
Comment (solegalli) [inline feature_engine/transformation/log.py:238]: **If auto,** this transformer learns the constant C to add to log transformation.

Could we add the above? otherwise it might be confusing?

File: feature_engine/transformation/log.py [hunk 13]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+
+        Raises
+        ------
+        TypeError
+            - If the input is not a Pandas DataFrame
+            - If any of the user provided variables are not numerical
+        ValueError
+            - If there are no numerical variables in the df or the df is empty
+            - If the variable(s) contain null values
+            - If some variables contain zero or negative values after suming constant C
```
Comment (solegalli) [inline feature_engine/transformation/log.py:261]: could we replace **suming** by **adding** please?

File: feature_engine/transformation/log.py [hunk 14]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+        # calculate C to add to each variable
+        if self.C == "auto":
+            self.C_ = X[self.variables_].min(axis=0).abs() + 1
+        else:
+            self.C_ = self.C
+
+        # check contains zero or negative values
+        if (X[self.variables_] + self.C_ <= 0).any().any():
+            raise ValueError(
+                "Some variables contain zero or negative values after suming"
```
Comment (solegalli) [inline feature_engine/transformation/log.py:280]: could we replace **suming** by **adding** please?

File: feature_engine/transformation/log.py [hunk 15]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+            The data to be transformed.
+
+        Raises
+        ------
+        TypeError
+            If the input is not a Pandas DataFrame
+        ValueError
+            - If the variable(s) contain null values
+            - If the df has different number of features than the df used in fit()
+            - If some variables contains zero or negative values after suming constant C
```
Comment (solegalli) [inline feature_engine/transformation/log.py:304]: could we replace **suming** by **adding** please?

File: feature_engine/transformation/log.py [hunk 16]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+                + "constant C, can't apply log"
+            )
+
+        self.n_features_in_ = X.shape[1]
+
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Transforms the variables using log transformation.
```
Comment (solegalli) [inline feature_engine/transformation/log.py:290]: Transforms the variables using **the** log transformation. Or more directly "Transforms the variables with the logarithm" (probably better).

Probably also needs changing in the old logtransformer above.

File: feature_engine/transformation/log.py [hunk 17]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = super().transform(X)
+
+        # check contains zero or negative values
+        if (X[self.variables_] + self.C_ <= 0).any().any():
+            raise ValueError(
+                "Some variables contain zero or negative values after suming"
```
Comment (solegalli) [inline feature_engine/transformation/log.py:318]: "adding"

File: feature_engine/transformation/log.py [hunk 18]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+        if self.base == "e":
+            X.loc[:, self.variables_] = np.log(X.loc[:, self.variables_] + self.C_)
+        elif self.base == "10":
+            X.loc[:, self.variables_] = np.log10(X.loc[:, self.variables_] + self.C_)
+
+        return X
+
+    def inverse_transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Transforms the variables using exponential transformation.
```
Comment (solegalli) [inline feature_engine/transformation/log.py:332]: We try to use the sklearn docstring: "Convert the data back to the original representation."

See for example here:
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html#sklearn.preprocessing.OrdinalEncoder.inverse_transform

File: feature_engine/transformation/log.py [hunk 19]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+    C_:
+        The constant C to transform data.
+
+    Methods
+    -------
+    fit:
+        This transformer learns the constant C to add to log transformation.
+    transform:
+        Transforms the variables using log transformation.
+    fit_transform:
```
Comment (solegalli) [inline feature_engine/transformation/log.py:215]: can we please add:
```
inverse_transform: 
      Convert the data back to the original representation.
```

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html#sklearn.preprocessing.OrdinalEncoder.inverse_transform

File: feature_engine/transformation/log.py [hunk 20]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+            The data to be detransformed.
+
+        Raises
+        ------
+        TypeError
+            If the input is not a Pandas DataFrame
+        ValueError
+            - If the variable(s) contain null values
+            - If X has different number of features than X used in fit()
+            - If some variables contains zero or negative values
```
Comment (solegalli) [inline feature_engine/transformation/log.py:346]: I don't think we should raise this error, because log of a value >0 and <1 will be negative, and can return the inverse. Am I correct?

File: feature_engine/transformation/log.py [hunk 21]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+        -------
+        X: Pandas dataframe
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = super().transform(X)
+
+        # check contains zero or negative values
+        if (X[self.variables_] <= 0).any().any():
```
Comment (solegalli) [inline feature_engine/transformation/log.py:358]: this check is suitable for transform, but not for inverse_transform, values can be negative.

This was probably in the PR i linked before, but we fixed it there already and we made a new PR with the updates. Sorry

File: tests/test_transformation/test_logcp_transformer.py [hunk 22]
```diff
@@ -0,0 +1,155 @@
+import pandas as pd
+import pytest
+from sklearn.exceptions import NotFittedError
+
+from feature_engine.transformation import LogCpTransformer
+
+_age_inverse_transform = [20.0, 21.0, 19.0, 18.0]
```
Comment (solegalli) [inline tests/test_transformation/test_logcp_transformer.py:7]: If we add one variable with values >0 and < than 1 we would catch the bug on having the test for <0 in the inverse_transform

File: tests/test_transformation/test_logcp_transformer.py [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+    # test error when data contains negative values
+    df_neg = df_vartypes.copy()
+    df_neg.loc[1, "Age"] = -df_vartypes["Age"].min() - 2
+
+    # when variable contains negative value, transform
+    with pytest.raises(ValueError):
+        transformer = LogCpTransformer()
+        transformer.fit(df_vartypes)
+        transformer.transform(df_neg)
+
```
Comment (solegalli) [inline tests/test_transformation/test_logcp_transformer.py:150]: do we need a test for when user specifies a c and the variable + the c is still negative during fit()? We have the error catch already, it is just a matter of adding a test.

I anticipate that many users will just make c=1 without even looking at the variable values, so transform seems a bit late to pick up this error.

**Review by solegalli — COMMENTED** [review]
@TremaMiguel 

Thank you so much for the quick turnaround and for such good and complete code. I love it :)

Just very tiny modifications requests to the docstrings and something that I believe could be a bug in the inverse_transform. And 1 or 2 small test addition requests :)

File: feature_engine/transformation/log.py [hunk 24]
```diff
@@ -164,3 +164,206 @@ def _more_tags(self):
...
+        -------
+        self
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+
+        # calculate C to add to each variable
+        if self.C == "auto":
+            self.C_ = X[self.variables_].min(axis=0).abs() + 1
```
Comment (solegalli) [inline feature_engine/transformation/log.py:273]: This is perhaps a painful request because it might slow things down, but I'll make it anyhow and we can discuss about it.

All Feature-engine transformers store learn parameters when they differ per variable as a dictionary, so that the user can easily see which value corresponds to each variable. And this is a key difference with sklearn arrays, where you can't really understand without some data manipulation what value corresponds to what.

I find this very useful. So I would like C_ to be a dictionary of variable: minimum +1 value pairs. Could we do that?

The display in rst would also need some changing after this modification

File: feature_engine/transformation/log.py (hunk 20 above)
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:346]: got it, additionally as the variable is already in log scale, the exponential function accepts negative and positive values, so there's no need for this error, thanks for catching it.

File: tests/test_transformation/test_logcp_transformer.py (hunk 22 above)
Comment (TremaMiguel) [reply tests/test_transformation/test_logcp_transformer.py:7]: ok, so if removing the check in inverse transform, no need for testing this, right?

File: tests/test_transformation/test_logcp_transformer.py (hunk 23 above)
Comment (TremaMiguel) [reply tests/test_transformation/test_logcp_transformer.py:150]: yes, good catch, when `C` is set manually then it is possible to raise this error. I'll add a test for that

File: feature_engine/transformation/log.py (hunk 24 above)
Comment (TremaMiguel) [reply feature_engine/transformation/log.py]: will add this also as input option for `C` param 

```python
C: "auto", int or dict, default="auto"
    The constant C to apply the transfomation log(x + C).

      - If int, then log(x + C)
      - If "auto", then C = abs(min(x)) + 1
      - If dict, dictionary mapping the constant C to apply
        to each variable.
```

**Comment by TremaMiguel** [comment]
@solegalli please take another look at changes

File: tests/test_transformation/test_logcp_transformer.py (hunk 22 above)
Comment (solegalli) [reply tests/test_transformation/test_logcp_transformer.py:7]: it doesn't hurt to test in my opinion :)

File: feature_engine/transformation/log.py (hunk 24 above)
Comment (solegalli) [reply feature_engine/transformation/log.py]: makes total sense, great idea

File: feature_engine/transformation/log.py (hunk 16 above)
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:296]: I've not seen this one, gonna make the change

File: tests/test_transformation/test_logcp_transformer.py (hunk 22 above)
Comment (TremaMiguel) [reply tests/test_transformation/test_logcp_transformer.py:7]: so if we add an "Score" variable here

```python
import pandas as pd
import numpy as np

data = {
        "Name": ["tom", "nick", "krish", "jack"],
        "City": ["London", "Manchester", "Liverpool", "Bristol"],
        "Age": [20, 21, 19, 18],
        "Marks": [0.9, 0.8, 0.7, 0.6],
        "Score": [0.15, 0.007, 0.32, 0.21],
        "dob": pd.date_range("2020-02-24", periods=4, freq="T"),
    }

df_vartypes = pd.DataFrame(data)
df_vartypes
```

what should we test for ?, considering we remove the check in `inverse_transform` 

```python 
# check contains zero or negative values
if (X[self.variables_] + self.C_ <= 0).any().any():
    raise ValueError(
        "Some variables contain zero or negative values after adding"
        + "constant C, can't apply log"
    )
```

Comment (solegalli) [reply tests/test_transformation/test_logcp_transformer.py:7]: You could use the "Marks" variable directly. That one should have catched the error raising the variables should not be negative in the inverse transform.

Comment (solegalli) [reply tests/test_transformation/test_logcp_transformer.py:7]: Oh, I understand the question now. Sorry. Applying the inverse_transform on a variable with values >-1, ensures the transformer works not only for variables >1 but also >0 and <1. And protects us automatically for introducing an error in the future. The error does not exist today, but we may accidentally introduce it in the future if we forget why we removed it, or if we modify some other masterclass from which this transformer inherits.

File: feature_engine/transformation/log.py [hunk 25]
```diff
@@ -164,3 +164,205 @@ def _more_tags(self):
...
+        if self.base == "e":
+            X.loc[:, self.variables_] = np.log(X.loc[:, self.variables_] + self.C_)
+        elif self.base == "10":
+            X.loc[:, self.variables_] = np.log10(X.loc[:, self.variables_] + self.C_)
+
+        return X
+
+    def inverse_transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Transforms the variables using the exponential transformation.
```
Comment (solegalli) [inline feature_engine/transformation/log.py:338]: The method should read "Convert the data back to the original representation." as per sklearn convention for inverse_transform.

File: tests/test_transformation/test_logcp_transformer.py (hunk 22 above)
Comment (TremaMiguel) [reply tests/test_transformation/test_logcp_transformer.py:7]: yes, thanks for clarifying, this is an easy addition, might be similar to `test_inverse_transform`

File: feature_engine/transformation/log.py (hunk 25 above)
Comment (TremaMiguel) [reply feature_engine/transformation/log.py:338]: got it

**Comment by TremaMiguel** [comment]
@solegalli added tests for float and int for constant `C`, think it is ready to be merged. Do you think we should change the name of the transformer ?

**Comment by solegalli** [comment]
> 
> 
> @solegalli added tests for float and int for constant `C`, think it is ready to be merged. Do you think we should change the name of the transformer ?

@TremaMiguel  thank you so much for the final changes! Yes, this is looking ready to merge.

Not sure about the name change, at the beginning, when I saw LogCpTransformer I did not quite understand what the p stands for. Only when I read the docs, I saw that the transformation is plus (p) a Constant (C). And then it made sense.

Ideally, we want the user to understand what the transformer does, directly from the transformer name. Not sure it this is possible in 100% of cases.

I think, I am happy with the name for the time being. Unless you have a suggestion?

PS: sorry for the delayed response, I am actually on holidays for the week :p

**Comment by TremaMiguel** [comment]
@solegalli ok ok, added your PR, let me know if anything else is needed.
